### PR TITLE
[aggregator] Log minor flush errors with readable metric context

### PR DESCRIPTION
**Pull request by olivielpeau**
### What does this PR do?

Adds readable metric context when logging a flush error (or an error that occurs when adding a sample).

### Motivation

In practice, changes the cryptic log messages like (with raw context keys):

```
2018-04-16 19:14:22 CEST | INFO | (context_metrics.go:76 in Flush) | An error occurred while flushing metric on context key '81cc203f3b14b19c8d780420187366fb': Rate value is negative, discarding it (the underlying counter may have been reset)
```

into messages with fully readable context elements:

```
2018-04-16 19:14:22 CEST | INFO | (check_sampler.go:44 in commit) | An error occurred while flushing check metric 'test.metric' on host 'myhost' and tags '[foo:bar foo:baz]': Rate value is negative, discarding it (the underlying counter may have been reset)
```

Should make it possible to understand these log messages.

### Additional Notes

Every commit can be reviewed independently. The implementation is a bit convoluted for what it does, happy to hear improvement suggestions.

### CODE REVIEW

File: pkg/aggregator/check_sampler.go [hunk 1]
```diff
@@ -32,11 +34,18 @@ func newCheckSampler(hostname string) *CheckSampler {
...
+		log.Debug("Ignoring sample '%s' on host '%s' and tags '%s': %s", metricSample.Name, metricSample.Host, metricSample.Tags, err)
+	}
 }
 
 func (cs *CheckSampler) commit(timestamp float64) {
-	for _, serie := range cs.metrics.Flush(timestamp) {
+	series, errors := cs.metrics.Flush(timestamp)
+	for ckey, err := range errors {
+		context := cs.contextResolver.contextsByKey[ckey]
+		log.Infof("An error occurred while flushing check metric '%s' on host '%s' and tags '%s': %s", context.Name, context.Host, context.Tags, err)
```
Comment (derekwbrown) [inline pkg/aggregator/check_sampler.go]: So, semantics here, but should it be `info` if we're logging an error? If it's non-fatal, then perhaps `warning`?

File: pkg/aggregator/check_sampler.go [hunk 2]
```diff
@@ -32,11 +34,18 @@ func newCheckSampler(hostname string) *CheckSampler {
...
+	if err := cs.metrics.AddSample(contextKey, metricSample, metricSample.Timestamp, 1); err != nil {
+		log.Debug("Ignoring sample '%s' on host '%s' and tags '%s': %s", metricSample.Name, metricSample.Host, metricSample.Tags, err)
+	}
 }
 
 func (cs *CheckSampler) commit(timestamp float64) {
-	for _, serie := range cs.metrics.Flush(timestamp) {
+	series, errors := cs.metrics.Flush(timestamp)
+	for ckey, err := range errors {
+		context := cs.contextResolver.contextsByKey[ckey]
```
Comment (derekwbrown) [inline pkg/aggregator/check_sampler.go]: check return value?  Otherwise log message below might crash.

File: pkg/aggregator/time_sampler.go [hunk 3]
```diff
@@ -139,6 +141,16 @@ func (s *TimeSampler) flush(timestamp float64) metrics.Series {
 	return result
 }
 
+// flushContextMetrics flushes the passed contextMetrics, handles its errors, and returns its series
+func (s *TimeSampler) flushContextMetrics(timestamp int64, contextMetrics metrics.ContextMetrics) []*metrics.Serie {
+	series, errors := contextMetrics.Flush(float64(timestamp))
+	for ckey, err := range errors {
+		context := s.contextResolver.contextsByKey[ckey]
+		log.Infof("An error occurred while flushing dogstatsd metric '%s' on host '%s' and tags '%s': %s", context.Name, context.Host, context.Tags, err)
```
Comment (derekwbrown) [inline pkg/aggregator/time_sampler.go]: Same

**Review by derekwbrown — DISMISSED** [review]

File: pkg/aggregator/check_sampler.go (hunk 2 above)
Comment (olivielpeau) [reply pkg/aggregator/check_sampler.go]: good point indeed. In no circumstances should the context key not be tracked by the `contextResolver` (unless something's horribly wrong in the aggregator's logic), that said we should still handle this case. Applies to the existing logic below too. I'll change this.

File: pkg/aggregator/check_sampler.go (hunk 1 above)
Comment (olivielpeau) [reply pkg/aggregator/check_sampler.go]: this should be an `info`, it's not really an error. I'll remove `error` from the message.

**Comment by olivielpeau** [comment]
@derekwbrown Thanks for the review! Addressed your comments, could you have another look?

**Review by arbll — APPROVED** [review]
LGTM

**Pull request merged**